core-plugins/src/test/java/io/cdap/plugin/batch/source/FTPBatchSourceTest.java

```diff
@@ -0,0 +1,164 @@
+import io.cdap.cdap.etl.mock.validation.MockFailureCollector;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.HashMap;
+
+/**
+ * Unit test for {@link FTPBatchSource.FTPBatchSourceConfig} class.
+ */
+
+public class FTPBatchSourceTest {
+
+  private static final String USER = "user";
+  private static final String PASSWORD_WITH_SPECIAL_CHARACTERS = "wex^Yz@123#456!";
+  private static final String PASSWORD_WITHOUT_SPECIAL_CHARACTERS = "wexYz123456";
+  private static final String HOST = "192.168.0.179";
+  private static final int FTP_DEFAULT_PORT = 21;
+  private static final int SFTP_DEFAULT_PORT = 22;
+  private static final String PATH = "/user-look-here.txt";
+  private static final String FTP_PREFIX = "ftp";
+  private static final String SFTP_PREFIX = "sftp";
+  private static final String SFTP_FS_CLASS = "org.apache.hadoop.fs.sftp.SFTPFileSystem";
+  private static final String FS_SFTP_IMPL = "fs.sftp.impl";
+
+  @Test
+  public void testFTPPathWithSpecialCharactersInAuth() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    config.configuration(String.format("%s://%s:%s@%s:%d%s", FTP_PREFIX, USER,
+                                       PASSWORD_WITH_SPECIAL_CHARACTERS, HOST, FTP_DEFAULT_PORT, PATH), null);
+    config.validate(collector);
+
+    final HashMap<String, String> fileSystemProperties = new HashMap<>();
+    fileSystemProperties.put("fs.ftp.host", HOST);
+    fileSystemProperties.put(String.format("fs.ftp.user.%s", HOST), USER);
+    fileSystemProperties.put(String.format("fs.ftp.password.%s", HOST), PASSWORD_WITH_SPECIAL_CHARACTERS);
+    fileSystemProperties.put("fs.ftp.host.port", String.valueOf(FTP_DEFAULT_PORT));
+    Assert.assertEquals(fileSystemProperties, config.getFileSystemProperties(collector));
+  }
+
+  @Test
+  public void testFTPPathWithoutSpecialCharactersInAuth() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    config.configuration(String.format("%s://%s:%s@%s:%d%s", FTP_PREFIX, USER, PASSWORD_WITHOUT_SPECIAL_CHARACTERS,
+                                       HOST, FTP_DEFAULT_PORT, PATH), null);
+    config.validate(collector);
+    Assert.assertEquals(config.getPathFromConfig(), config.getPath());
+    Assert.assertEquals(0, config.getFileSystemProperties(collector).size());
+  }
+
+  @Test
+  public void testMissingAuthInPath() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    config.configuration(String.format("%s://%s:%d%s", FTP_PREFIX, HOST, FTP_DEFAULT_PORT, PATH), null);
+    try {
+      config.validate(collector);
+    } catch (ValidationException e) {
+      Assert.assertEquals(1, e.getFailures().size());
```

Comment:
Validation will always throw an exception in the case when authentication is missing. 

This is the line of code that will throw the exception in this case:

https://github.com/cdapio/hydrator-plugins/blob/182830262f7c6f2aee85ebec21c19105b6f6b6ba/core-plugins/src/main/java/io/cdap/plugin/batch/source/FTPBatchSource.java#L314

Reply:
The problem is that if someone introduces a bug where they fail to throw an exception when authentication is missing, your test won't catch it. That's why we need to verify that:
1. an exception was thrown -> missing this check
2. the exception thrown is the one that you expected -> you're already checking this

Reply:
You're right. Now I got your point. I was missing Assert.fail() in try {} block. This will make sure that an exception will be thrown when authentication is missing.